consoles/sshVirtsh.pm

```diff
@@ -325,111 +325,169 @@ sub add_interface {
+and destination directory. If successful, returns absolute path to downloaded file
+
+    $self->download_image_file('/var/lib/openqa/share/factory/hdd/SUSE-MicroOS.s390x-5.1.0-Default-kvm-Build10.9.qcow2', '/var/lib/libvirt/images/');
+
+=cut
+
+sub download_image_file {
+    my ($self, $src, $dest) = @_;
+    $dest .= basename($src);
+    die "Downloading $src failed!" if $self->run_cmd("rsync -av $src $dest");
+    return $dest;
+}
+
+sub decompress_cmd {
+    return sprintf(q(xz --decompress --keep --verbose --no-warn --quiet '%s'), shift);
+}
+
 sub add_disk {
     my ($self, $args) = @_;
+    die('`backingfile` or `cdrom` has been used without a given image file') if (($args->{backingfile} || $args->{cdrom}) && !$args->{file});
+    die('Please specify device id (dev_id)') unless ($args->{dev_id});
 
-    my $backingfile             = $args->{backingfile};
-    my $cdrom                   = $args->{cdrom};
-    my $name                    = $self->name;
-    my $file                    = $name . $args->{dev_id} . ($self->vmm_family eq 'vmware' ? '.vmdk' : '.img');
-    my $basedir                 = '/var/lib/libvirt/images/';
-    my $vmware_datastore        = get_var('VMWARE_DATASTORE', '');
-    my $vmware_openqa_datastore = "/vmfs/volumes/$vmware_datastore/openQA/";
-    if ($args->{create}) {
-        my $size = $args->{size} || '20G';
-        if ($self->vmm_family eq 'vmware') {
-            my $vmware_disk_path = $vmware_openqa_datastore . $file;
-            # Power VM off, delete it's disk image, and create it again.
-            # Than wait for some time for the VM to *really* turn off.
-            my $cmd =
-              "( set -x; vmid=\$(vim-cmd vmsvc/getallvms | awk \'/$name/ { print \$1 }\');" .
-              'if [ $vmid ]; then ' .
-              'vim-cmd vmsvc/power.off $vmid;' .
-              'vim-cmd vmsvc/destroy $vmid;' .
-              'fi;' .
-              "vmkfstools -v1 -U $vmware_disk_path;" .
-              "vmkfstools -v1 -c $size --diskformat thin $vmware_disk_path; sleep 10 ) 2>&1";
-            my $retval = $self->run_cmd($cmd, domain => 'sshVMwareServer');
-            die "Can't create VMware image $vmware_disk_path" if $retval;
+    my $size = $args->{size} // '20G';
+    $size .= 'G' if ($size =~ /^\d+$/);
+
+    my $basedir;
+    my $file;
+    my $vmware_datastore = get_var('VMWARE_DATASTORE', '');
+
+    if ($self->vmm_family eq 'vmware') {
+        my $basedir     = "/vmfs/volumes/$vmware_datastore/openQA/";
+        my $nfs_basedir = '/vmfs/volumes/' . get_required_var('VMWARE_NFS_DATASTORE');
+        $file = sprintf("%s%s.vmdk", $self->{name}, $args->{dev_id});
+        my $vmware_disk_path = $basedir . $file;
+
+        # Power VM off, delete it's disk image, and create it again.
+        # Than wait for some time for the VM to *really* turn off.
+        my $cmd = <<"EOT";
+( set -x; vmid=\$(vim-cmd vmsvc/getallvms | awk '/$self->{name}/ { print \$1 }');
+if [ \$vmid ]; then
```

Comment:
Is it actually required to escape the `$` here?

Reply:
Yes, that should be treated as a shell variable.

Reply:
You're right, apparently interpolation is happening here. (`<<'EOT'` would avoid it but it looks like you actually need it here.)